.github/workflows/backport-queue.yml

```diff
@@ -0,0 +1,169 @@
+  contents: read
+
+jobs:
+  commitQueue:
+    runs-on: ubuntu-latest
+    if: github.repository == 'nodejs/node' || github.event_name == 'workflow_dispatch'
+    strategy:
+      fail-fast: false
+      matrix:
+        branch:
+          - { release-line: v23.x, base-branch: main }
+          - { release-line: v22.x, base-branch: v23.x }
+    steps:
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683  # v4.2.2
+        with:
+          # Needs the whole git history for ncu to work
+          # See https://github.com/nodejs/node-core-utils/pull/486
+          fetch-depth: 0
+          ref: ${{ matrix.branch.release-line }}-staging
+          # A personal token is required because pushing changes to `.github`
+          # with default token will be blocked. It needs
+          # to be set here because `checkout` configures GitHub authentication
+          # for push as well.
+          token: ${{ secrets.GH_USER_TOKEN }}
+
+      - name: Set up Python ${{ env.PYTHON_VERSION }}
+        uses: actions/setup-python@0b93645e9fea7318ecaed2b359559ac225c90a2b  # v5.3.0
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      # Install dependencies
+      - name: Install Node.js
+        uses: actions/setup-node@39370e3970a6d050c480ffad4ff0ed4d3fdee5af  # v4.1.0
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+      - name: Install branch-diff
+        run: npm install -g branch-diff
+
+      - name: Setup git author
+        run: |
+          git config --local user.email "github-bot@iojs.org"
+          git config --local user.name "Node.js GitHub Bot"
+
+      - name: Set up ghauth config (Ubuntu)
+        run: |
+          mkdir -p "${XDG_CONFIG_HOME:-~/.config}/changelog-maker"
+          echo '{}' | jq '{user: env.USERNAME, token: env.TOKEN}' > "${XDG_CONFIG_HOME:-~/.config}/changelog-maker/config.json"
+        env:
+          USERNAME: ${{ secrets.JENKINS_USER }}
+          TOKEN: ${{ github.token }}
+
+      - name: Fetch base branch
+        run: |
+          git remote set-branches --add origin "$BASE_BRANCH"
+          git fetch origin "$BASE_BRANCH"
+        env:
+          BASE_BRANCH: ${{ matrix.branch.base-branch }}
+
+      - name: Fetch auto-backport branch if it exists
+        id: fetch
+        run: |
+          if git fetch origin "$BACKPORT_BRANCH"; then
+            STAGING_BRANCH_TIP="$(git rev-parse HEAD)"
+            WORKING_BRANCH_TIP="$(git rev-parse FETCH_HEAD)"
+            echo "WORKING_BRANCH_TIP=$WORKING_BRANCH_TIP" >> "$GITHUB_OUTPUT"
+            if [ "$WORKING_BRANCH_TIP" != "$STAGING_BRANCH_TIP" ]; then
+                git reset  "$WORKING_BRANCH_TIP" --hard
+                git rebase "$STAGING_BRANCH_TIP" --empty=drop || git reset "$STAGING_BRANCH_TIP" --hard
+            fi
+          else
+            echo "Branch doesn't exist yet"
+          fi
+        env:
+          BACKPORT_BRANCH: ${{ matrix.branch.release-line }}-staging-auto-backport
```

Comment:
We have to keep in mind that the purpose of the `v<MAJOR>.x-staging` branches is precisely to serve as a working area for the releaser, in the sense that the releaser is free to break builds, force-push, and modify it however they want when preparing a release.

With that in mind, adding a second `<MAJOR>-staging-auto-backport` seems like an unnecessary duplication of effort that I'm not sure is adding value here.

Reply:
I addressed that point in the PR description, we cannot use the commits backported by the automation, so it makes more sense to use a different branch IMO. It could also be on a separate repo if we don't want to clutter nodejs/node with that

Reply:
ok, let me copy/paste each point so that we can discuss each one in more detail:

> It's only able to work with commits that cleanly apply to the staging branch, which is trivial for a human to do.

It's not trivial and _very_ time consuming, particularly because these "trivial to cherry-pick" commits are mixed up with many conflicting commits that may or not be simple for the releaser to handle it themselves. This automation should be precisely about separating this two types of commits, whatever lands cleanly can go in the staging branch and whatever needs backport can be handled later - later, as in the next 10mins of work for that releaser or throughout the week, it's gives more control to the releaser on how they want to handle it.

> No matter how small of a conflict, as soon as a commit doesn't backport it is skipped, and having out of order backports makes branches diverge more and more, making future backport harder.

Agreed, that's why the report generated by https://github.com/nodejs/node-core-utils/pull/875 is in the correct order so that the releaser can do a best-effort work to try and land the easy-to-manage conflicted commits in the right order. Any non-trivial conflict will still require the backport from (likely) the original PR author as usual.

> Releasers sign their backport commits. Because commits that land on staging branch receive much fewer attention than the one on main, we don't want anyone to be tempted to introduce subtle backdoor when backporting, having the signature puts the releaser reputation at stake. I don't think we could have the same system with a bot while making sure no one can impersonate the bot.

I think you're talking about the manual backport process when you mention "we don't want anyone to be tempted to introduce subtle backdoor when backporting" - that, has its own process, which consists in opening a new PR that still needs to be reviewed manually (and approved at least by the releaser) and which we usually run tests again, in order to land in the staging branch.

I'm failing to see how the bot impersonation plays out in that case.

Reply:
> I'm failing to see how the bot impersonation plays out in that case.

If the bot is pushing commits on the staging branch, it would be very easy for a malicious actor to impersonate the bot and push a commit that contains e.g. a backdoor. It's not a problem on `main` which has a lot of eyes on it, but on staging branches, it would be very hard to spot, and possibly tricky to track down. Currently, the way we have things where only humans are landing stuff on the staging branches puts a good incentive not to be malicious actor, and to be careful not to leak our private keys, as the blame would go to whoever signed the malicious commit.


> that, has its own process

I think it would be silly to assume a malicious actor would follow process 🙃